vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc

```diff
@@ -724,6 +731,38 @@ void AdsImpl::OnPageLoaded(
       << previous_tab_url_;
 }
 
+void AdsImpl::ExtractPurchaseIntentSignal(const std::string& url) {
+  if (!ShouldClassifyPagesIfTargeted()) {
+    return;
+  }
+
+  if (!TestSearchState(url) &&
+      helper::Uri::MatchesDomainOrHost(url, previous_tab_url_)) {
+    return;
+  }
+
+  PurchaseIntentSignalInfo purchase_intent_signal =
+      purchase_intent_classifier_->ExtractIntentSignal(url);
+
+  if (purchase_intent_signal.segments.empty() &&
+      purchase_intent_signal.timestamp_in_seconds == 0) {
+    return;
+  }
+
+  GeneratePurchaseIntentSignalHistoryEntry(purchase_intent_signal);
+  BLOG(INFO) << "Purchase intent signal extracted for " << url;
+}
+
+void AdsImpl::GeneratePurchaseIntentSignalHistoryEntry(
+    const PurchaseIntentSignalInfo& purchase_intent_signal) {
+  for (const auto& segment : purchase_intent_signal.segments) {
+    auto history = std::make_unique<PurchaseIntentSignalHistory>();
```

Comment:
`PurchaseIntentSignalHistory history` as then we do not need to dereference below